### PR TITLE
Add ElementIndex

### PR DESCRIPTION
**Pull request by wthrowe**

### CODE REVIEW

File: src/Domain/ElementIndex.hpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  ElementIndex_UNUSED
+  unsigned block_id_ : block_id_bits;
+  ElementIndex_UNUSED
+  unsigned refinement_level_ : refinement_bits;
+  ElementIndex_UNUSED
+  unsigned index_ : max_refinement_level;
+};
+}  // namespace ElementIndex_detail
+
+/// A class for indexing a Charm array by Element.
```
Comment (nilsdeppe) [inline src/Domain/ElementIndex.hpp]: We should probably add this to the `Parallel` doxygen group

File: src/Domain/ElementIndex.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace ElementIndex_detail {
+SegmentIndex::SegmentIndex(size_t block_id,
+                           const SegmentId& segment_id) noexcept
+  : block_id_(block_id), refinement_level_(segment_id.refinement_level()),
+    index_(segment_id.index()) {
+  ASSERT(block_id < two_to_the(block_id_bits),
+         "Block id out of bouunds: " << block_id);
```
Comment (nilsdeppe) [inline src/Domain/ElementIndex.cpp]: bouunds -> bounds

**Review by kidder — COMMENTED** [review]
fix and squash

**Comment by wthrowe** [comment]
In addition to the requested changes I adjusted the refinement level assertion to check that it was <= max_refinement_level instead of just that it fit in the allocated bits and modified the associated message to print the out-of-range value before it was truncated to the bitfield.

**Review by kidder — APPROVED** [review]

**Pull request merged**